File: Chapter 05/C#/Section_5_4_Server_Monitor/After/Infrastructure/Adapters/KafkaAlertAdapter.cs
using Chapter05.ServerMonitor.After.Core.Ports;
using Chapter05.ServerMonitor.After.Infrastructure.ExternalLibs;
using System;

namespace Chapter05.ServerMonitor.After.Infrastructure.Adapters
{
    /// <summary>
    /// ADAPTER 3: The "Scale" Adapter (Async Messaging).
    /// Handles the transformation of domain alerts into JSON messages for Kafka.
    /// </summary>
    public class KafkaAlertAdapter : IAlertPort
    {
        private readonly IProducer<string, string> _producer;

        /// <summary>
        /// Injects a Messaging Producer into the adapter.
        /// </summary>
        public KafkaAlertAdapter(IProducer<string, string> producer)
        {
            _producer = producer;
        }

        /// <summary>
        /// Formats and pushes the alert to an asynchronous message broker.
        /// </summary>
        public void SendAlert(string message)
        {
            // The Adapter is responsible for formatting data for the outside world.
            string payload = $"{{\"Error\": \"{message}\", \"Timestamp\": \"{DateTime.UtcNow}\"}}";
            
            // We use a static key to ensure chronological ordering within a single partition.
            _producer.Produce("Server-01", "server-alerts", payload);
        }
    }
}

File: Chapter 05/Java/src/main/java/com/grokkingsoftwarearchitecture/chapter05/section_5_4_server_monitor/after/infrastructure/adapters/KafkaAlertAdapter.java
package com.grokkingsoftwarearchitecture.chapter05.section_5_4_server_monitor.after.infrastructure.adapters;

import com.grokkingsoftwarearchitecture.chapter05.section_5_4_server_monitor.after.core.ports.AlertPort;
import com.grokkingsoftwarearchitecture.chapter05.section_5_4_server_monitor.after.infrastructure.externallibs.Producer;
import java.time.Instant;

/**
 * ADAPTER 3: The "Scale" Adapter (Async Messaging).
 * Wraps a messaging producer to allow the system to scale via a message broker.
 */
public class KafkaAlertAdapter implements AlertPort {
    private final Producer<String, String> kafkaProducer;

    public KafkaAlertAdapter(Producer<String, String> kafkaProducer) {
        this.kafkaProducer = kafkaProducer;
    }

    @Override
    public void sendAlert(String message) {
        String payload = String.format("{\"Error\": \"%s\", \"Timestamp\": \"%s\"}", 
                                        message, Instant.now());
        
        // Passing a Key ("Server-01") ensures message order for this specific server.
        kafkaProducer.produce("Server-01", "server-alerts-topic", payload);
    }
}